app/services/person_constructor.rb

```diff
@@ -0,0 +1,82 @@
+class PersonConstructor
+  KEY_NAME_MAPPINGS = {
+    address: :address_1,
+    zip: :zip_code,
+  }
+  OLD_REMOTE_PARAMS = [
+    remote_fields: PersonWithRemoteFields::REMOTE_PARAMS
+  ]
+
+  def self.permitted_params
+    Person::PERMITTED_PARAMS +
+      PersonWithRemoteFields::REMOTE_PARAMS +
+      Location::PERMITTED_PARAMS +
+      KEY_NAME_MAPPINGS.keys +
+      OLD_REMOTE_PARAMS +
+      [:uuid] # this should come from PersonFinder
+  end
+
+  def initialize(params)
+    @params = params.deep_symbolize_keys
+    normalize_params
+  end
+
+  def build
+    person = find_or_initialize_person_with_remote_fields
+    person.assign_attributes(person_params)
+    assign_location_attributes(person)
+    person
+  end
+
+  private
+
+  attr_reader :params
+
+  def find_or_initialize_person_with_remote_fields
+    if person = PersonFinder.new(params).find
+      person.becomes(PersonWithRemoteFields)
+    else
+      PersonWithRemoteFields.new
+    end
+  end
+
+  def assign_location_attributes(person)
+    if location_params.any?
+      new_location = LocationComparable.new(location_params)
+      person.location.becomes(LocationComparable).merge(new_location)
+      person.location.fill_in_missing_attributes
+    end
+  end
+
+  def normalize_params
+    flatten_remote_fields
+    rename_keys
+    strip_whitespace_from_values
+  end
+
+  def flatten_remote_fields
+    params.merge!(params.delete(:remote_fields) || {})
+  end
+
+  def strip_whitespace_from_values
+    params.merge!(params){ |k, v1| v1.try(:strip) || v1 }
+  end
+
+  def rename_keys
```

Comment:
also, I was hoping that `normalize_keys` and `flatten_remote_fields` would
both be temporary. Ideally, the frontend should be passing in the correct
params so we don't have to worry about mapping

On Mon, Mar 7, 2016 at 6:41 PM, Mario Olivio Flores <
notifications@github.com> wrote:

> In app/services/person_constructor.rb
> <https://github.com/MayOneUS/mayday-2.0-backend/pull/86#discussion_r55293682>
> :
>
> > +
> > +  def normalize_params
> > +    flatten_remote_fields
> > +    rename_keys
> > +    strip_whitespace_from_values
> > +  end
> > +
> > +  def flatten_remote_fields
> > +    params.merge!(params.delete(:remote_fields) || {})
> > +  end
> > +
> > +  def strip_whitespace_from_values
> > +    params.merge!(params){ |k, v1| v1.try(:strip) || v1 }
> > +  end
> > +
> > +  def rename_keys
>
> normalize_keys makes sense to me! Thanks!
> … <#-1948800572_>
> On 3/7/2016 3:39 PM, Jacob Turino wrote: > + rename_keys
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/MayOneUS/mayday-2.0-backend/pull/86/files#r55293682>.
>